hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go

```diff
@@ -184,6 +185,134 @@ func (builder *convertToARMBuilder) referencePropertyHandler(
 	)
 }
 
+func (builder *convertToARMBuilder) flattenedPropertyHandler(
+	toProp *astmodel.PropertyDefinition,
+	fromType *astmodel.ObjectType) []dst.Stmt {
+
+	toPropName := toProp.PropertyName()
+
+	// collect any fromProps that were flattened from the to-prop
+	var fromProps []*astmodel.PropertyDefinition
+	for _, prop := range fromType.Properties() {
+		if prop.WasFlattenedFrom(toPropName) {
+			fromProps = append(fromProps, prop)
+		}
+	}
+
+	// there are none to copy; exit
+	if len(fromProps) == 0 {
+		return nil
+	}
+
+	allTypes := builder.codeGenerationContext.GetAllReachableTypes()
+
+	// the toProp shape here must be:
+	// 1. maybe a typename, pointing to…
+	// 2. maybe optional, wrapping …
+	// 3. maybe a typename, pointing to…
+	// 4. an object type
+
+	// (1.) resolve the outer typename
+	toPropType, err := allTypes.FullyResolve(toProp.PropertyType())
+	if err != nil {
+		panic(err)
+	}
+
+	needToInitializeToProp := false // we need to init the target if it is optional
+	var toPropTypeName astmodel.TypeName
+	// (2.)  resolve any optional type
+	if optType, ok := toPropType.(*astmodel.OptionalType); ok {
+		needToInitializeToProp = true
+		// (3.) resolve any inner typename
+		toPropTypeName = optType.Element().(astmodel.TypeName)
+		toPropType, err = allTypes.FullyResolve(optType.Element())
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// (4.) we have found the underlying object type
+	toPropObjType := toPropType.(*astmodel.ObjectType)
```

Comment:
Same question, if this part would be clearer as a method that was just called here?

Wondering further: Is this and the code in the other direction actually the same (modulo setting the `needToInitializeToProp`)? Maybe you could refactor to a shared function: `func lookupFlattenedType(t astmodel.Type) (*astmodel.ObjectType, wasOptional bool, error)`?

Reply:
I did try this at first but it didn't come out very clean, the methods are a bit lopsided. I can try again when I look to redo it a bit to support multiple levels of flattening, but I'd rather push to get the PR done now.